Add Windows CI

joefitzgerald opened the pull request:
## Done
- [x] Add Windows CI via Appveyor
- [x] Pave the road for releases to be generated from Appveyor (Windows)
- [x] Make various improvements to script/cibuild and build/Gruntfile.coffee
- [x] Add caching of `~/.atom/.npm`, `~/.atom/.apm`, and `~/.atom/.node-gyp` on Appveyor
## Future
- [ ] Pave the road for releases to be generated from Travis (OSX / Linux)
- [ ] Provide a pathway for Node v4.2.1 / apm 1.5.x, but do not make them the default for internal Atom CI
- [ ] Add caching of `~/.atom/.npm`, `~/.atom/.apm`, and `~/.atom/.node-gyp` on Travis
- [ ] Make remoteReleases synchronization faster
- [ ] Add caching of electron
- [ ] Add caching of node in `https://github.com/atom/apm/blob/master/script/download-node.js`
- [ ] Create zip archive for Windows
- [ ] Add artifacts explicitly instead of all of `atom-build`
- [ ] Disable `0.10.35` builds on Windows (we now know they work)
- [ ] Add `signtool` tool
- [ ] Add encrypted ATOM_ACCESS_TOKEN to .travis.yml, with scopes necessary for release upload
- [ ] Fix .deb package generation
- [ ] Add encrypted signing keys for Travis and Appveyor, with corresponding encrypted variable

50Wliu commented on script/cibuild:
```diff
@@ -22,16 +22,18 @@ function loadEnvironmentVariables(filePath) {
-    process.env.CC = 'clang';
-    process.env.CXX = 'clang++';
-    process.env.npm_config_clang = '1';
+  if (process.env.JANKY_SHA1) {
```
:sparkles: Was thinking of creating a PR for this to get rid of the ENOENT errors on Travis.

nathansobo commented:
Nice. Appreciate the task list for status.

50Wliu commented on .travis.yml:
```diff
@@ -14,23 +14,32 @@ compiler: clang
+  - yes | cp build/package-node-$NODE_VERSION.json build/package.json
 
-script: script/cibuild
+script: script/cibuild | ts
```
`ts` doesn't seem to exist on Mac.  Is this even needed anyway?  All it does is add a timestamp which I don't think is particularly useful for CI.

50Wliu commented:
This would also probably be a good time to switch Travis's language to `node_js`.  Right now it's unset so it's defaulting to Ruby, which really isn't helping at all.  Switching to `node_js` should allow us to get rid of the manual nvm installation/Node configuration steps and let Travis handle it for us.

```yaml
language: node_js
node_js:
  - "4.2.1" # or alternatively, "stable"
  - "0.10.35" # 0.10 should work as well
```

50Wliu commented:
Appveyor also seems to support directly setting Node versions, though slightly differently from Travis: http://www.appveyor.com/docs/lang/nodejs-iojs

50Wliu commented on build/tasks/set-version-task.coffee:
```diff
@@ -5,10 +5,15 @@ module.exports = (grunt) ->
   getVersion = (callback) ->
-    onBuildMachine = process.env.JANKY_SHA1 and process.env.JANKY_BRANCH in ['stable', 'beta']
+    releasableBranches = ['stable', 'beta']
+    inCI = process.env.CI or process.env.JANKY_SHA1
```
Doesn't seem to be used

joefitzgerald replied:
Fixed in b4555a167a48bbea7ae6a5d0de07f9cf2a0edb82.

joefitzgerald commented:
> Appveyor also seems to support directly setting Node versions, though slightly differently from Travis: http://www.appveyor.com/docs/lang/nodejs-iojs

Already using the technique described.

joefitzgerald commented:
> This would also probably be a good time to switch Travis's language to node_js.

Done.

joefitzgerald commented:
@50Wliu I have to back out use of the `language: node_js` support; it's not available on OSX yet: https://github.com/travis-ci/travis-ci/issues/2311.

50Wliu commented:
> it's not available on OSX yet: travis-ci/travis-ci#2311.

Ugh, whyyy

50Wliu commented on build/Gruntfile-ci.coffee:
```diff
@@ -0,0 +1,295 @@
+  grunt.registerTask('lint', ['standard', 'coffeelint', 'csslint', 'lesslint'])
+  grunt.registerTask('test', ['shell:kill-atom', 'run-specs'])
+
+  ciTasks = ['output-disk-space', 'download-electron', 'download-electron-chromedriver', 'build']
```
`output-disk-space` should be disabled for at least Travis, and maybe Appveyor.  It currently doesn't output anything on Travis.

joefitzgerald replied:
Fixed in 6edb96ff9fc9af78b12356b357a914771fbd9790.

nathansobo commented:
Kicked off a build on our internal CI infra so we can be sure we're still working there, too.

joefitzgerald committed:
Add Appveyor CI

- Add Appveyor badge
- Add Appveyor.yml definition, with encrypted ATOM_ACCESS_TOKEN
- Add cibuild.cmd script to launch cibuild on Windows
- Prepare to release Atom on Windows from Appveyor (but don't do it yet)
- Reduce dependence on JANKY in ci scripts + tasks
- Suppress release activities on PRs for Travis and Appveyor
- Cache ~/.atom/.apm, ~/.atom/.node-gyp/.atom, and ~/.atom/.npm on Appveyor

nathansobo commented:
Kicked off another build after the force-push. Sorry, should have done this yesterday.

nathansobo commented:
:zap: Thanks for your patience on this, both with the build and with us. Very excited about moving to more commodified build infra.

maxbrunsfeld commented:
:clap: :moneybag: :chart_with_upwards_trend:

mnquintana commented:
:rocket: :guitar: :clap:
